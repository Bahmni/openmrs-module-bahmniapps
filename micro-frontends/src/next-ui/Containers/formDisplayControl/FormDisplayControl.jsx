import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { Accordion, AccordionItem } from "carbon-components-react";
import "../../../styles/carbon-conflict-fixes.scss";
import "../../../styles/carbon-theme.scss";
import "../../../styles/common.scss";
import "./formDisplayControl.scss";
import { FormattedMessage } from "react-intl";
import {
  fetchFormData,
  getLatestPublishedForms,
} from "../../utils/FormDisplayControl/FormUtils";
import {
  buildFormMap,
  findByEncounterUuid,
  doesUserHaveAccessToTheForm,
} from "../../utils/FormDisplayControl/FormView";
import { I18nProvider } from "../../Components/i18n/I18nProvider";
import ViewObservationForm from "../../Components/ViewObservationForm/ViewObservationForm";
import { formatDate } from "../../utils/utils";
import EditObservationForm from "../../Components/EditObservationForm/EditObservationForm";

/** NOTE: for reasons known only to react2angular,
 * any functions passed in as props will be undefined at the start, even ones inside other objects
 * so you need to use the conditional operator like props.hostApi?.callback even though it is a mandatory prop
 */

export function FormDisplayControl(props) {
  const noFormText = (
    <FormattedMessage
      id={"NO_FORM"}
      defaultMessage={"No Form found for this patient...."}
    />
  );
  const formsHeading = (
    <FormattedMessage
      id={"DASHBOARD_TITLE_FORMS_2_DISPLAY_CONTROL_KEY"}
      defaultMessage={"Observation Forms"}
    />
  );
  const loadingMessage = (
    <FormattedMessage
      id={"LOADING_MESSAGE"}
      defaultMessage={"Loading... Please Wait"}
    />
  );

  const [formList, setFormList] = useState([]);
  const [isLoading, setLoading] = useState(true);
  const [showViewObservationForm, setViewObservationForm] = useState(false);
  const [showEditObservationForm, setEditObservationForm] = useState(false);
  const [formName, setFormName] = useState("");
  const [formNameTranslations, setFormNameTranslations] = useState("");
  const [formData, setFormData] = useState([]);
  const [isViewFormLoading, setViewFormLoading] = useState(false);
  const [isEditFormLoading, setEditFormLoading] = useState(false);
  const [encounterUuid, setEncounterUuid] = useState("");

  const buildResponseData = async () => {
    try {
      const formResponseData = await fetchFormData(
        props?.hostData?.patientUuid,
        props?.hostData?.numberOfVisits
      );
      const latestForms = await getLatestPublishedForms();
      var grouped = {};
      if (formResponseData?.length > 0 && latestForms?.length > 0) {
        formResponseData.forEach(function (formEntry) {
          const latestForm = latestForms.find(
            (latestForm) => formEntry.formName == latestForm.name
          );
          grouped[formEntry.formName] = grouped[formEntry.formName] || [];
          grouped[formEntry.formName].push({
            encounterDate: formEntry.encounterDateTime,
            encounterUuid: formEntry.encounterUuid,
            visitUuid: formEntry.visitUuid,
            visitDate: formEntry.visitStartDateTime,
            providerName: formEntry.providers[0].providerName,
            providerUuid: formEntry.providers[0].uuid,
            formVersion: formEntry.formVersion,
            formNameTranslations: getFormDisplayName(latestForm),
            privileges: latestForm?.privileges,
          });
        });
      }
      Object.keys(grouped).forEach(function (key) {
        grouped[key] = grouped[key].sort((a, b) => {
          return new Date(b.encounterDate) - new Date(a.encounterDate);
        });
      });
      setFormList(grouped);
    } catch (e) {
      console.log(e);
    } finally {
      setLoading(false);
    }
  };

  const getFormDisplayName = function (form) {
    const locale = localStorage.getItem("NG_TRANSLATE_LANG_KEY") || "en";
    const formNameTranslations = form?.nameTranslation
      ? JSON.parse(form.nameTranslation)
      : [];
    let formDisplayName = form?.name;
    if (formNameTranslations.length > 0) {
      const currentLabel = formNameTranslations.find(function (
        formNameTranslation
      ) {
        return formNameTranslation.locale === locale;
      });
      formDisplayName = currentLabel ? currentLabel.display : form?.name;
    }
    return formDisplayName;
  };

  const showEdit = function (currentEncounterUuid) {
    return props?.hostData?.showEditForActiveEncounter
      ? props?.hostData?.encounterUuid === currentEncounterUuid
      : true;
  };

  const handleEditSave = (encounter) => {
    props?.hostApi?.handleEditSave(encounter);
  };

  const openViewObservationForm = async (
    formName,
    encounterUuid,
    formNameTranslations
  ) => {
    var formMap = {
      formName: formName,
      encounterUuid: encounterUuid,
      hasNoHierarchy: props?.hostData?.hasNoHierarchy,
    };
    setFormName(formName);
    setFormNameTranslations(formNameTranslations);
    setViewFormLoading(true);
    setViewObservationForm(true);
    const data = await buildFormMap(formMap);
    setViewFormLoading(false);
    setFormData(data[0].value[0].groupMembers);
  };

  const openEditObservationForm = async (
    formName,
    encounterUuid,
    formNameTranslations
  ) => {
    var formMap = {
      formName: formName,
      encounterUuid: encounterUuid,
      hasNoHierarchy: props?.hostData?.hasNoHierarchy,
    };
    setEditFormLoading(true);
    setEditObservationForm(true);
    const data = await findByEncounterUuid(formMap.encounterUuid);
    const filteredFormObs = data.observations.filter((obs) =>
      obs.formFieldPath?.includes(formName)
    );
    setFormName(formName);
    setFormNameTranslations(formNameTranslations);
    setEncounterUuid(encounterUuid);
    setFormData(filteredFormObs);
  };

  const closeViewObservationForm = () => {
    setFormData([]);
    setFormName("");
    setFormNameTranslations("");
    setViewObservationForm(false);
  };
  const closeEditObservationForm = () => {
    setFormData([]);
    setFormName("");
    setFormNameTranslations("");
    setEditObservationForm(false);
  };

  const checkForPrivileges = (data, action) => {
    return doesUserHaveAccessToTheForm(
      props?.hostData?.currentUser?.privileges,
      data,
      action
    );
  };

  const printForm = () => {
    formData.map(function(obs) {
      if (obs?.groupMembers?.length > 0) {
        obs.encounterDateTime = obs?.groupMembers[0].encounterDateTime;
      }
    })
    props?.hostApi?.printForm(formData);
  };

  useEffect(() => {
    buildResponseData();
  }, []);

  return (
    <>
      <I18nProvider>
        <div>
          <h2 className={"forms-display-control-section-title"}>
            {formsHeading}
          </h2>
          {isLoading ? (
            <div className="loading-message">{loadingMessage}</div>
          ) : (
            <div className={"placeholder-text-forms-control"}>
              {Object.entries(formList).length > 0
                ? Object.entries(formList).map(([key, value]) => {
                    const moreThanOneEntry = value.length > 1;
                    return moreThanOneEntry ? (
                      <Accordion>
                        <AccordionItem
                          title={value[0].formNameTranslations}
                          className={"form-accordion"}
                          open
                        >
                          {value.map((entry, index) => {
                            return (
                              <div key={index} className={"row-accordion"}>
                                <span className={"form-name-text"}>
                                  {checkForPrivileges(entry, "view") ? (
                                    <a
                                      onClick={() =>
                                        openViewObservationForm(
                                          key,
                                          entry.encounterUuid,
                                          entry.formNameTranslations
                                        )
                                      }
                                      className="form-link"
                                    >
                                      {formatDate(entry.encounterDate)}
                                    </a>
                                  ) : (
                                    formatDate(entry.encounterDate)
                                  )}
                                  {checkForPrivileges(entry, "edit") &&
                                    showEdit(entry.encounterUuid) && (
                                      <i
                                        className="fa fa-pencil"
                                        onClick={() => {
                                          openEditObservationForm(
                                            key,
                                            entry.encounterUuid,
                                            entry.formNameTranslations
                                          );
                                        }}
                                      ></i>
                                    )}
                                </span>
                                <span className={"form-provider-text"}>
                                  {entry.providerName}
                                </span>
                              </div>
                            );
                          })}
                        </AccordionItem>
                      </Accordion>
                    ) : (
                      <div className={"form-display-control-row"}>
                        <span
                          className={
                            "form-non-accordion-text form-heading form-name"
                          }
                        >
                          {value[0].formNameTranslations}
                        </span>
                        <span
                          className={
                            "form-non-accordion-text form-date-align form-date-time"
                          }
                        >
                          {checkForPrivileges(value[0], "view") ? (
                            <a
                              className="form-link"
                              onClick={() =>
                                openViewObservationForm(
                                  key,
                                  value[0].encounterUuid,
                                  value[0].formNameTranslations
                                )
                              }
                            >
                              {formatDate(value[0].encounterDate)}
                            </a>
                          ) : (
                            formatDate(value[0].encounterDate)
                          )}
                          {checkForPrivileges(value[0], "edit") &&
                            showEdit(value[0].encounterUuid) && (
                              <i
                                className="fa fa-pencil"
                                onClick={() => {
                                  openEditObservationForm(
                                    key,
                                    value[0].encounterUuid,
                                    value[0].formNameTranslations
                                  );
                                }}
                              ></i>
                            )}
                        </span>
                        <span
                          className={"form-non-accordion-text provider-name"}
                        >
                          {value[0].providerName}
                        </span>
                      </div>
                    );
                  })
                : noFormText}
              {showViewObservationForm ? (
                <ViewObservationForm
                  isViewFormLoading={isViewFormLoading}
                  formName={formName}
                  formNameTranslations={formNameTranslations}
                  closeViewObservationForm={closeViewObservationForm}
                  formData={formData}
                  showPrintOption={props?.hostData?.showPrintOption}
                  printForm={printForm}
                />
              ) : null}
              {showEditObservationForm ? (
                <EditObservationForm
                  isEditFormLoading={isEditFormLoading}
                  formName={formName}
                  formNameTranslations={formNameTranslations}
                  closeEditObservationForm={closeEditObservationForm}
                  patient={props?.hostData?.patient}
                  formData={formData != [] && formData}
                  encounterUuid={encounterUuid}
                  consultationMapper={props?.hostData?.consultationMapper}
                  handleEditSave={handleEditSave}
                  setEditFormLoading={setEditFormLoading}
                  editErrorMessage={props?.hostData?.editErrorMessage}
                />
              ) : null}
            </div>
          )}
        </div>
      </I18nProvider>
    </>
  );
}

FormDisplayControl.propTypes = {
  hostData: PropTypes.object.isRequired,
  hostApi: PropTypes.object.isRequired
};
