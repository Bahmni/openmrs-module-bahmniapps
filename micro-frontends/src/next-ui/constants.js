export const defaultDateFormat = "DD MMM YYYY";
export const defaultDateTimeFormat = "DD MMM YYYY hh:mm a";

export const LS_LANG_KEY = "NG_TRANSLATE_LANG_KEY";
export const MEDICATION_ACKNOWLEDGE_SQL_PROPERTY = "emrapi.sqlSearch.emergencyMedicationToAcknowledge";
export const BASE_URL =
  typeof __webpack_public_path__ !== "undefined"
    ? __webpack_public_path__
    : "/";

const hostUrl = localStorage.getItem("host")
  ? "https://" + localStorage.getItem("host")
  : "";

export const verifierFunction = "Verifier";
export const document_images_path = "../../document_images/";
const RESTWS_V1 = hostUrl + "/openmrs/ws/rest/v1";

export const FORM_BASE_URL = RESTWS_V1 + "/bahmnicore/patient/{patientUuid}/forms";
export const ENCOUNTER_BASE_URL = RESTWS_V1 + "/bahmnicore/bahmniencounter/{encounterUuid}";
export const GET_ALL_FORMS_BASE_URL = RESTWS_V1 + "/bahmniie/form/allForms";
export const GET_FORMS_BASE_URL = RESTWS_V1 + "/form/{formUuid}";
export const GET_FORM_TRANSLATE_URL = RESTWS_V1 + "/bahmniie/form/translate";
export const FETCH_CONCEPT_URL = RESTWS_V1 + "/concept/{conceptUuid}?v=full&locale={locale}";
export const BAHMNI_ENCOUNTER_URL = RESTWS_V1 + "/bahmnicore/bahmniencounter";
export const ENCOUNTER_TYPE_URL = RESTWS_V1 + "/encountertype/{encounterType}";
export const GET_ALLERGIES_URL = "/openmrs/ws/fhir2/R4/AllergyIntolerance?patient={patientId}&_summary=data"
export const FORM_TRANSLATIONS_URL =  RESTWS_V1 + "/bahmniie/form/translations";
export const OBSERVATIONS_URL = RESTWS_V1 + "bahmnicore/observations";
export const LATEST_PUBLISHED_FORMS_URL = RESTWS_V1 + "/bahmniie/form/latestPublishedForms";
export const GET_DRUG_ACKNOWLEDGEMENT_URL = RESTWS_V1 + "/bahmnicore/sql?q={property}&v=full&location_uuid={location_uuid}&provider_uuid={provider_uuid}";
export const EMERGENCY_MEDICATIONS_BASE_URL = RESTWS_V1 + "/ipd/adhocMedicationAdministrations/{medication_administration_uuid}";
export const GET_PROVIDER_UUID_URL = RESTWS_V1 + "/session";
export const SAVE_ALLERGIES_URL =  RESTWS_V1 + "/patient/{patientId}/allergy";
export const OT_NOTES_BASE_URL = RESTWS_V1 + "/notes";

