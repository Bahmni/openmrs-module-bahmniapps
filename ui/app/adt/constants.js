'use strict';

var Bahmni = Bahmni || {};
Bahmni.ADT = Bahmni.ADT || {};

Bahmni.ADT.Constants = (function () {
    return {
        patientsListUrl: "/patient/search",
        ipdDashboard: "#/patient/{{patientUuid}}/visit/{{visitUuid}}/",
        admissionLocationUrl: "/openmrs/ws/rest/v1/admissionLocation/"
    };
})();

