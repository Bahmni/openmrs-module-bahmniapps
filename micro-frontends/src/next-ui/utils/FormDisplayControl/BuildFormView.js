import axios from "axios";
import { GET_ALL_FORMS_BASE_URL, GET_FORMS_BASE_URL, GET_FORM_TRANSLATE_URL } from "../../constants";
import { cloneDeep, remove } from "lodash";


var getAllForms = async () => {
    const apiURL = GET_ALL_FORMS_BASE_URL;
    const params = {
        v: "custom:(version,name,uuid)"
    };
    try {
        const response = await axios.get(apiURL, { params });
        if (response.status === 200) {
            return response.data;
        }
        return [];
    } catch (error) {
        console.error(error);
    }
};

var getFormDetail = async (formUuid ) => {
    const apiURL = GET_FORMS_BASE_URL.replace('{formUuid}', formUuid);
    const params = { 
        v: "custom:(resources:(value))"
    };
    try {
        const response = await axios.get(apiURL, { params });
        if (response.status === 200) {
            return response.data;
        }
        return [];
    } catch (error) {
        console.error(error);
    }
};


var getFormTranslate = async (formName, formVersion, locale, formUuid ) => {
    const apiURL = GET_FORM_TRANSLATE_URL;
    const params = {
        "formName": formName,
        "formVersion": formVersion,
        "locale": locale,
        "formUuid": formUuid
    };
    try {
        const response = await axios.get(apiURL, { params });
        if (response.status === 200) {
            return response.data;
        }
        return [];
    } catch (error) {
        console.error(error);
    }
};
var formBuildForms = [];
export const build = async (bahmniObservations, hasNoHierarchy) => {
    bahmniObservations.forEach(function (obs) {
        obs.value = preProcessMultiSelectObs(obs.value);
    });
    formBuildForms = await getAllForms();
    var obs = createObsGroupForForm(bahmniObservations, formBuildForms);
    if (!hasNoHierarchy) {
        const newObs = await updateObservationsWithFormDefinition(obs, formBuildForms);
        return newObs;
    }
};

var getLatestObservationDateTime = function (groupMembers) {
    var latestObservationDateTime = groupMembers[0].observationDateTime;
    groupMembers.forEach(function (member) {
        latestObservationDateTime = latestObservationDateTime < member.observationDateTime ? member.observationDateTime : latestObservationDateTime;
    });
    return latestObservationDateTime;
};

var createMultiSelectObservation = function (observations) {
    var multiSelectObject = {};
    multiSelectObject.type = "multiSelect";
    multiSelectObject.concept = observations[0].concept;
    multiSelectObject.encounterDateTime = observations[0].encounterDateTime;
    // multiSelectObject.groupMembers = observations; // might need to add observations based on other approaches.
    multiSelectObject.groupMembers = [];
    const multiSelectValue = observations.map(obs => obs.value && obs.value.shortName || obs.value)
    multiSelectObject.value = multiSelectValue.join(', ');
    multiSelectObject.conceptConfig = {multiSelect: true};
    multiSelectObject.observationDateTime = getLatestObservationDateTime(observations);
    multiSelectObject.providers = observations[0].providers;
    multiSelectObject.creatorName = observations[0].creatorName;
    multiSelectObject.formFieldPath = observations[0].formFieldPath;
    multiSelectObject.encounterUuid = observations[0].encounterUuid;
    return multiSelectObject;
};

var preProcessMultiSelectObs = function (value) {
    var clonedGroupMembers = cloneDeep(value);
    clonedGroupMembers.forEach( function (member) {
        if (member && member.groupMembers.length === 0) {
            var obsWithSameFormFieldPath = getRecordObservations(member.formFieldPath, value);
            if (obsWithSameFormFieldPath.length > 1) {
                var multiSelectObject = createMultiSelectObservation(obsWithSameFormFieldPath);
                value.push(multiSelectObject);
            } else if (obsWithSameFormFieldPath.length === 1) {
                value.push(obsWithSameFormFieldPath[0]);
            }
        } else if (member.groupMembers.length > 0) {
            var obsGroups = getRecordObservations(member.formFieldPath, value);
            obsGroups.forEach( function (obsGroup) {
                obsGroup.groupMembers = preProcessMultiSelectObs(obsGroup.groupMembers);
                value.push(obsGroup);
            });
        }
    });
    return value;
};

var createObsGroupForForm = function (observations, formBuilderForms) {
    observations.forEach( function (obs) {
        var newValues = [];
        obs.value.forEach( function (value) {
            if (!value.formFieldPath) {
                newValues.push(value);
                return;
            }
            var obsGroup = {
                "groupMembers": [],
                "concept": {
                    "shortName": "",
                    "conceptClass": null
                },
                "encounterUuid": ""

            };
            var formName = value.formFieldPath.split('.')[0];
            var formBuilderForm = formBuilderForms.find(function (form) {
                return form.name ===
                    formName;
            });
            obsGroup.concept.shortName = formName;
            var locale = localStorage.getItem("NG_TRANSLATE_LANG_KEY") || "en";
            var formNameTranslations = formBuilderForm && formBuilderForm.nameTranslation ? JSON.parse(formBuilderForm.nameTranslation) : [];
            if (formNameTranslations.length > 0) {
                var currentLabel = formNameTranslations.find(function (formNameTranslation) {
                        return formNameTranslation.locale === locale;
                    });
                if (currentLabel) {
                    obsGroup.concept.shortName = currentLabel.display;
                }
            }
            obsGroup.encounterUuid = value.encounterUuid;
            var previousObsGroupFound;
            newValues.forEach( function (newValue) {
                if (obsGroup.concept.shortName === newValue.concept.shortName) {
                    newValue.groupMembers.push(value);
                    previousObsGroupFound = true;
                }
            });
            if (previousObsGroupFound) {
                return;
            }
            obsGroup.groupMembers.push(value);
            newValues.push(obsGroup);
        });
        obs.value = newValues;
    });
    return observations;
};

var updateObservationsWithFormDefinition = async function (observations, formBuildForms) {
    var allForms = formBuildForms;
    for(let observation of observations) {
        var forms = [];
        for (let form of observation.value) {
            if (form.concept.conceptClass) {
                forms.push(form);
                return;
            }
            var observationForm = getFormByFormName(allForms, getFormName(form.groupMembers), getFormVersion(form.groupMembers));
            if (!observationForm) {
                return;
            }
            var response = await getFormDetail(observationForm.uuid)
            //doubt: which resource it should pick - response.resources[0].value;
            var formDetailsAsString = response.resources[0].value;
            if (formDetailsAsString) {
                var formDef = JSON.parse(formDetailsAsString);
                formDef.version = observationForm.version;
                var locale = localStorage["NG_TRANSLATE_LANG_KEY"] || "en";
                var translationData = await getFormTranslate(formDef.name, formDef.version, locale, formDef.uuid)
                forms.push(updateObservationsWithRecordTree(formDef, form, translationData));
                observation.value = forms;
            }
            observation.value = forms;
        };
    };
    return observations;
};

var getFormByFormName = function (formList, formName, formVersion) {
    return formList.find( function (form) {
        return form.name === formName && form.version === formVersion;
    });
};

var getFormName = function (members) {
    var member = members.find( function (member) {
        return member.formFieldPath !== null;
    });
    return member ? member.formFieldPath.split('.')[0] : undefined;
};

var getFormVersion = function (members) {
    var member = members.find( function (member) {
        return member.formFieldPath !== null;
    });
    return member ? member.formFieldPath.split('.')[1].split('/')[0] : undefined;
};

var updateObservationsWithRecordTree = function (formDef, form, translationData) {
    var recordTree = getRecordTree(formDef, form.groupMembers);
    recordTree = JSON.parse(JSON.stringify(recordTree));
    createGroupMembers(recordTree, form, form.groupMembers, translationData);
    return form;
};

var createColumnGroupsForTable = function (record, columns, tableGroup, obsList, translationData) {
    columns.forEach( function (column, index) {
        var obsGroup = {
            "groupMembers": [],
            "concept": {
                "shortName": "",
                "conceptClass": null
            }
        };
        var translationKey = column.translationKey;
        var defaultShortName = column.value;
        obsGroup.concept.shortName = getTranslatedShortName(translationData, translationKey, obsGroup, defaultShortName);
        var columnRecord = getColumnObs(index, record);
        column.children = columnRecord;
        createGroupMembers(column, obsGroup, obsList, translationData);
        if (obsGroup.groupMembers.length > 0) {
            tableGroup.groupMembers.push(obsGroup);
        }
    });
};

var getTranslatedShortName = function (translationData, translationKey, obsGroup, defaultShortName) {
    if (isTranslationKeyPresent(translationData, translationKey)) {
        return translationData.labels[translationKey][0];
    }
    return defaultShortName;
};

var isTranslationKeyPresent = function (translationData, translationKey) {
    return translationData && translationData.labels &&
        translationData.labels[translationKey][0] !== translationKey;
};

var getColumnObs = function (columnIndex, record) {
    var columnChildren = [];
    record.children.map( function (child) {
        if (child.control.properties && child.control.properties.location.column === columnIndex) {
            columnChildren.push(child);
        }
    });
    return columnChildren;
};

var createGroupMembers = function (recordTree, obsGroup, obsList, translationData) {
    recordTree.children.forEach( function (record) {
        if (record.control.type === 'obsControl' || record.control.type === 'obsGroupControl') {
            var recordObservations = getRecordObservations(record.formFieldPath, obsList);
            recordObservations.forEach( function (recordObservation) {
                obsGroup.groupMembers.push(recordObservation);
            });
        }
        else if (record.control.type === 'section') {
            var sectionGroup = createObsGroup(record, translationData);
            createGroupMembers(record, sectionGroup, obsList, translationData);
            if (sectionGroup.groupMembers.length > 0) {
                obsGroup.groupMembers.push(sectionGroup);
            }
        }
        else if (record.control.type === "table") {
            var tableGroup = createObsGroup(record, translationData);
            var columns = record.control.columnHeaders;
            createColumnGroupsForTable(record, columns, tableGroup, obsList, translationData);
            if (tableGroup.groupMembers.length > 0) {
                obsGroup.groupMembers.push(tableGroup);
            }
        }
    });
};

var getRecordObservations = function (obsFormFieldPath, obsList) {
    return remove(obsList, function (obs) {
        return obs.formFieldPath && obs.formFieldPath === obsFormFieldPath;
    });
};

var createObsGroup = function (record, translationData) {
    var obsGroup = {
        "groupMembers": [],
        "concept": {
            "shortName": "",
            "conceptClass": null
        }
    };
    var translationKey = record.control.label.translationKey;
    var defaultShortName = record.control.label.value;
    obsGroup.concept.shortName = getTranslatedShortName(translationData, translationKey, obsGroup, defaultShortName);
    return obsGroup;
};