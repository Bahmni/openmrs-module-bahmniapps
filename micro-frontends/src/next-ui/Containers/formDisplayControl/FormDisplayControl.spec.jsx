import React from "react";
import moment from "moment";
import { render, screen, waitFor } from "@testing-library/react";
import { FormDisplayControl } from "./FormDisplayControl";
import {
  mockFormResponseData,
  mockLatestPublishedForms,
  mockFormResponseDataForPrivilege,
  mockLatestPublishedFormsWithEditPrivileges,
  mockLatestPublishedFormsWithViewPrivileges,
  mockLatestPublishedFormsWithBothViewEditPrivileges,
  mockLatestPublishedFormsWithoutBothViewEditPrivileges,
} from "./FormDisplayControlMockData";
import { defaultDateTimeFormat } from "../../constants";

const mockFetchFormData = jest.fn();
const mockGetLatestPublishedForms = jest.fn();

jest.mock("../../utils/FormDisplayControl/FormUtils", () => ({
  fetchFormData: () => mockFetchFormData(),
  getLatestPublishedForms: () => mockGetLatestPublishedForms(),
}));

jest.mock("../../Components/i18n/I18nProvider", () => ({
  I18nProvider: ({ children }) => <div>{children}</div>,
}));

const mockHostData = {
  patientUuid: "some-patient-uuid",
  showEditForActiveEncounter: true,
  encounterUuid: "some-encounter-uuid",
};

const activeEncounterMockHostDataWithPrivileges = {
  patientUuid: "some-patient-uuid",
  showEditForActiveEncounter: true,
  encounterUuid: "6e52cecd-a095-457f-9515-38cf9178cb50",
  currentUser: {
    privileges: [
      {
        name: "View/Edit Forms",
        retired: false,
      },
    ],
  },
};

describe("FormDisplayControl Component for empty mock data", () => {
  it("should show no-forms-message when form entries are empty", async () => {
    const mockWithPatientHostData = {
      patientUuid: "some-patient-uuid",
      encounterUuid: undefined,
    };
    mockFetchFormData.mockResolvedValueOnce({});

    const { container } = render(
      <FormDisplayControl hostData={mockWithPatientHostData} />
    );

    await waitFor(() => {
      expect(
        screen.getByText("No Form found for this patient....")
      ).toBeTruthy();
      expect(
        container.querySelector(".placeholder-text-forms-control").innerHTML
      ).toEqual("No Form found for this patient....");
    });
  });
});

describe("FormDisplayControl Component", () => {
  it("should render the component", () => {
    const { container } = render(
      <FormDisplayControl hostData={mockHostData} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should show loading message", () => {
    const { container } = render(
      <FormDisplayControl hostData={mockHostData} />
    );
    expect(container.querySelector(".loading-message")).not.toBeNull();
    expect(container.querySelector(".loading-message").innerHTML).toEqual(
      "Loading... Please Wait"
    );
  });
});

describe("FormDisplayControl Component with Accordion and Non-Accordion", () => {
  beforeEach(() => {
    mockFetchFormData.mockResolvedValue(mockFormResponseData);
    mockGetLatestPublishedForms.mockResolvedValue(mockLatestPublishedForms);
  });
  // TODO: fix this test
  //  it("should render the component with form data", async() => {
  //   mockFetchFormData.mockResolvedValueOnce(mockFormResponseData);
  //   const { container } = render(<FormDisplayControl hostData={mockHostData} />);
  //   await waitFor(() => {
  //     expect(container).toMatchSnapshot();
  //   });
  // });

  it("should render accordion form entries when loading is done", async () => {
    const { container } = render(
      <FormDisplayControl hostData={mockHostData} />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".bx--accordion__title")).toHaveLength(
        1
      );
      expect(
        container.querySelector(".bx--accordion__title").innerHTML
      ).toEqual("Pre Anaesthesia Assessment");
      expect(
        container.querySelector(".row-accordion > .form-name-text > .form-link")
          .innerHTML
      ).toEqual(moment(1693217959000).format(defaultDateTimeFormat));
      expect(
        container.querySelector(".row-accordion > .form-provider-text")
          .innerHTML
      ).toEqual("Doctor One");
    });
  });

  it("should render non-accordion form entries when loading is done", async () => {
    const { container } = render(
      <FormDisplayControl hostData={mockHostData} />
    );

    await waitFor(() => {
      expect(
        container.querySelectorAll(".form-non-accordion-text")
      ).toHaveLength(6);
      expect(
        container.querySelectorAll(".form-non-accordion-text.form-heading")[0]
          .innerHTML
      ).toEqual("Orthopaedic Triage");
      expect(
        container.querySelectorAll(
          ".form-non-accordion-text.form-date-align > a"
        )[0].innerHTML
      ).toEqual(moment(1693277657000).format(defaultDateTimeFormat));
      expect(
        container.querySelectorAll(".form-non-accordion-text")[2].innerHTML
      ).toEqual("Doctor Two");
      expect(
        container.querySelectorAll(".form-non-accordion-text.form-heading")[1]
          .innerHTML
      ).toEqual("Patient Progress Notes and Orders");
      expect(
        container.querySelectorAll(
          ".form-non-accordion-text.form-date-align > a"
        )[1].innerHTML
      ).toEqual(moment(1693277657000).format(defaultDateTimeFormat));
      expect(
        container.querySelectorAll(".form-non-accordion-text")[5].innerHTML
      ).toEqual("Doctor One");
    });
  });

  it("should not see edit button for non-active-encounter entries and when showEditForActiveEncounter is true", async () => {
    const { container } = render(
      <FormDisplayControl hostData={mockHostData} />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(0);
    });
  });

  it("should see edit button for active-encounter entries and when showEditForActiveEncounter is true", async () => {
    const activeEncounterMockHostData = {
      patientUuid: "some-patient-uuid",
      showEditForActiveEncounter: true,
      encounterUuid: "6e52cecd-a095-457f-9515-38cf9178cb50",
    };
    const { container } = render(
      <FormDisplayControl hostData={activeEncounterMockHostData} />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(2);
    });
  });

  it("should see edit button for all entries and when showEditForActiveEncounter is false", async () => {
    const activeEncounterMockHostData = {
      patientUuid: "some-patient-uuid",
      showEditForActiveEncounter: false,
      encounterUuid: "6e52cecd-a095-457f-9515-38cf9178cb50",
    };
    const { container } = render(
      <FormDisplayControl hostData={activeEncounterMockHostData} />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(4);
    });
  });

  it("should see edit button for all entries and when showEditForActiveEncounter is not present", async () => {
    const activeEncounterMockHostData = {
      patientUuid: "some-patient-uuid",
      encounterUuid: "6e52cecd-a095-457f-9515-38cf9178cb50",
    };
    const { container } = render(
      <FormDisplayControl hostData={activeEncounterMockHostData} />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(4);
    });
  });

  it("should see edit button based on privilege", async () => {
    mockFetchFormData.mockResolvedValue(mockFormResponseDataForPrivilege);
    mockGetLatestPublishedForms.mockResolvedValue(
      mockLatestPublishedFormsWithEditPrivileges
    );

    const { container } = render(
      <FormDisplayControl
        hostData={activeEncounterMockHostDataWithPrivileges}
      />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".form-link")).toHaveLength(0);
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(1);
    });
  });

  it("should see view button based on privilege", async () => {
    mockFetchFormData.mockResolvedValue(mockFormResponseDataForPrivilege);
    mockGetLatestPublishedForms.mockResolvedValue(
      mockLatestPublishedFormsWithViewPrivileges
    );

    const { container } = render(
      <FormDisplayControl
        hostData={activeEncounterMockHostDataWithPrivileges}
      />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".form-link")).toHaveLength(1);
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(0);
    });
  });

  it("should see view button based on privilege", async () => {
    mockFetchFormData.mockResolvedValue(mockFormResponseDataForPrivilege);
    mockGetLatestPublishedForms.mockResolvedValue(
      mockLatestPublishedFormsWithBothViewEditPrivileges
    );

    const { container } = render(
      <FormDisplayControl
        hostData={activeEncounterMockHostDataWithPrivileges}
      />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".form-link")).toHaveLength(1);
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(1);
    });
  });

  it("should see view button based on privilege", async () => {
    mockFetchFormData.mockResolvedValue(mockFormResponseDataForPrivilege);
    mockGetLatestPublishedForms.mockResolvedValue(
      mockLatestPublishedFormsWithoutBothViewEditPrivileges
    );

    const { container } = render(
      <FormDisplayControl
        hostData={activeEncounterMockHostDataWithPrivileges}
      />
    );

    await waitFor(() => {
      expect(container.querySelectorAll(".form-link")).toHaveLength(0);
      expect(container.querySelectorAll(".fa.fa-pencil")).toHaveLength(0);
    });
  });
});
