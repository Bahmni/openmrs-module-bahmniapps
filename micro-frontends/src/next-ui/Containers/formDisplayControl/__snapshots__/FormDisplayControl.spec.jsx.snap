// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDisplayControl Component should render the component 1`] = `
<div>
  <div>
    <div>
      <h2
        class="forms-display-control-section-title"
      >
        Observation Forms
      </h2>
      <div
        class="loading-message"
      >
        Loading... Please Wait
      </div>
    </div>
  </div>
</div>
`;
